CDCgov/prime-reportstream
# experience/5713: new validate endpoint

This PR adds a new endpoint at `/api/validate` that returns any errors or warnings that a normal submission would, but without logging, writing to the database, sending to blob storage or (of course) sending to a receiver. The meat of the code was written by @clediggins-usds, with Experience adding kdoc, unit tests and some miscellaneous gentle refactors (but with Experience owning the endeavor (of course)). The endpoint currently supports Okta authentication (a la the waters endpoint) and only "topic" senders for now (no "full ELR" for this pass).

Test Steps:
1. Shape an API call in curl or Postman based on an `/api/waters` call, but replace "waters" with "validate"
2. Compare the errors and warnings JSON in the returned responses from `/api/waters` and `/api/validate` - they should match

## Changes
- Sub-class of SubmissionReceiver called ValidationReceiver created
- New entry point /api/validate in new file ValidationFunction.kt created
- New validateClaim() method extracted from submitToWaters endpoint and added to AuthenticationStrategy.kt to be dryly shared with /api/validate. Refactored the funky negative logic in this method while we were there.
- Added kdoc and unit tests as needed to all of the above

## Checklist

### Testing

## Linked Issues
- Relates to #5713 

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
